Content.Shared/Actions/SharedActionsSystem.cs

```diff
@@ -0,0 +1,396 @@
+
+        _popupSystem.PopupEntity(msg, performer, filter);
+
+        return true;
+    }
+    #endregion
+
+    #region AddRemoveActions
+    /// <summary>
+    ///     Add actions to an action component.
+    /// </summary>
+    /// <param name="uid">Entity to receive the actions</param>
+    /// <param name="actions">The actions</param>
+    /// <param name="provider">The entity that enables these actions (e.g., flashlight). May be null (innate actions).</param>
+    public virtual void AddActions(EntityUid uid, IEnumerable<ActionType> actions, EntityUid? provider, ActionsComponent? comp = null, bool dirty = true)
+    {
+        if (!Resolve(uid, ref comp, false))
+            return;
+
+        foreach (var action in actions)
+        {
+            action.Provider = provider;
+            action.AttachedEntity = comp.Owner;
+        }
+
+        // Sometimes the client receives actions from the server, before predicting that newly added components will add their own shared actions.
+        comp.Actions.ExceptWith(actions);
+        comp.Actions.UnionWith(actions);
+
+        if (dirty)
+            Dirty(comp);
+    }
+
+    public void RemoveProvidedActions(EntityUid uid, EntityUid provider, ActionsComponent? comp = null)
+    {
+        if (!Resolve(uid, ref comp, false))
+            return;
+
+        var provided = comp.Actions.Where(act => act.Provider == provider).ToList();
+
+        RemoveActions(uid, provided, comp);
+    }
+
+    public virtual void RemoveActions(EntityUid uid, IEnumerable<ActionType> actions, ActionsComponent? comp = null, bool dirty = true)
+    {
+        if (!Resolve(uid, ref comp, false))
+            return;
+
+        foreach (var action in actions)
+        {
+            comp.Actions.Remove(action);
+            action.AttachedEntity = null;
+        }
+
+        if (dirty)
+            Dirty(comp);
+    }
+
+    public void RemoveAction(EntityUid uid, ActionType action, ActionsComponent? comp = null)
+        => RemoveActions(uid, new[] { action }, comp);
+
+    public void AddAction(EntityUid uid, ActionType action, EntityUid? provider, ActionsComponent? comp = null)
+        => AddActions(uid, new[] { action }, provider, comp);
+    #endregion
+
+    #region EquipHandlers
+    private void OnDidEquip(EntityUid uid, ActionsComponent component, DidEquipEvent args)
+    {
+        var ev = new GetActionsEvent();
+        RaiseLocalEvent(args.Equipment, ev, false);
+
+        if (ev.Actions.Count == 0)
+            return;
+
+        AddActions(args.Equipee, ev.Actions, args.Equipment, component);
+    }
+
+    private void OnHandEquipped(EntityUid uid, ActionsComponent component, DidEquipHandEvent args)
+    {
+        var ev = new GetActionsEvent();
+        RaiseLocalEvent(args.Equipped, ev, false);
+
+        if (ev.Actions.Count == 0)
+            return;
+
+        AddActions(args.User, ev.Actions, args.Equipped, component);
+    }
+    #endregion
```

Comment:
doesnt unequip need to be handled or am i just missing that part

Reply:
The event handler for those was just a lambda function, but I'll make them "real" functions.
```cs
SubscribeLocalEvent<ActionsComponent, DidUnequipHandEvent>((uid, comp, args) => RemoveProvidedActions(uid, args.Unequipped, comp));
SubscribeLocalEvent<ActionsComponent, DidUnequipEvent>((uid, comp, args) => RemoveProvidedActions(uid, args.Equipment, comp));
```